RFC005: Distributed Network over gRPC

reinkrul opened the pull request:
Impact to reference implementation:

Required to be functionally compliant:
- Update broadcasting protocol to use blocks
- Don't accept transactions older than T-2

For the rest:
- Make separate issues for security measures

woutslakhorst reviewed (changes requested):
We need to decide on what to put in this RFC and what not. A lot of the threads can be discussed at a later point. It's good to have an overview somewhere and when an issue is solved, it's translated to RFCs. Maybe make a thread project board?

reinkrul commented on rfc/rfc005-distributed-network-using-grpc.md:
```diff
@@ -0,0 +1,259 @@
+
+1. The local node broadcasts their DAG state at a set interval:
+   - the references of the current day's (`T`) DAG head transactions.
+   - the block (see "Blocks") hashes of the previous 3 days (`T-1`, `T-2`, `T-3`). The hash of `T-3` is special because it also includes
```
Changed the description a bit, since it's irrelevant whether it's calculated from the past blocks or transactions: both should lead to the same result since it's a XOR operation.

reinkrul commented on rfc/rfc005-distributed-network-using-grpc.md:
```diff
@@ -0,0 +1,259 @@
+adds more load to a system possibly under stress. A back-off strategy SHOULD be used which only reconnects after an
+every increasing waiting period.
+
+In future versions we can add the concept of "dead nodes", but due to the anticipated size of the network it's not
```
will make an issue

reinkrul commented on rfc/rfc005-distributed-network-using-grpc.md:
```diff
@@ -0,0 +1,259 @@
+
+### 5.3. Querying Peer's DAG
+
+When the local node decides to query a peer's DAG because it differs from its own, it uses the `TransactionListQuery` message.
```
I hope not?

reinkrul replied:
Ah I see I forgot to spec that you have to specify the block;

```
// TransactionListQuery is a message used to query a peer's TransactionList.
message TransactionListQuery {
    // blockDate specifies which block to query the transactions for.
    uint32 blockDate = 1;
}
```

woutslakhorst replied:
If T-3 is one "block" then it could...

reinkrul replied:
The idea is that if you query T-3 you'll get only that block's transaction. It's just that when you broadcast it, the last hash contains all hashes up until and including T-3. I'll clarify it more.

reinkrul commented on rfc/rfc005-distributed-network-using-grpc.md:
```diff
@@ -0,0 +1,259 @@
+by the node operator but SHALL NOT be so short that it clutters the network. It is advised to keep it relatively short
+because it directly influences the speed by which new transactions are propagated to peers.  
+
+### 5.3. Querying Peer's DAG
```
added

reinkrul commented on rfc/rfc005-distributed-network-using-grpc.md:
```diff
@@ -0,0 +1,259 @@
+A (malicious) peer could exhaust the node's memory with (many) large messages.
+
+Countermeasures:
+- Nodes SHOULD NOT accept incoming messages larger than 5mb (5242880 bytes).
```
Good question, "incoming messages" concerns both transactions and payload (since they's both network messages). I'd say both?

reinkrul commented on rfc/rfc005-distributed-network-using-grpc.md:
```diff
@@ -0,0 +1,259 @@
+steal identities (e.g. DIDs).
+
+Countermeasures:
+- Nodes MUST verify the payload hash when receiving transaction payload.
```
Yes, RFC004. Added the ref.

woutslakhorst commented on rfc/rfc005-distributed-network-using-grpc.md:
```diff
@@ -0,0 +1,259 @@
+In this protocol each day is a block. Every day at midnight (UTC timezone) a new block starts.
+The signing time of a transaction determines which block it belongs to. As such the signing time of the DAG's root transaction determines the first block.
+
+### 5.1.1. Block Hash
```
The T-3 block computation will get heavier each day

reinkrul replied:
> The T-3 block computation will get heavier each day

Which is fine, as long as you don't recalculate all historical blocks every time you need `T-3`. It's a simple XOR operation, so I expect when spinning up a new node the bottleneck will be actually fetching and verifying the DAG (from a peer), XOR-ing all refs is one of the faster operations as part of that I think.

woutslakhorst replied:
According to this PR, T-3 is calculated by sorting all hashes (not per block), which means that every day the entire history needs to be resorted.

woutslakhorst reviewed (commented):
The protocol still has a problem when a node publishes TXs in T-2. The clock drift will differ at nodes, the result will be that one node will accept the TX when it still falls within T-2 when another node thinks it falls within T-3 and will reject it.

reinkrul commented:
> The protocol still has a problem when a node publishes TXs in T-2 and they are received around 00:00:00. The clock drift will differ at nodes, the result will be that one node will accept the TX when it still falls within T-2 when another node thinks it falls within T-3 and will reject it.

The tx's `signingTime` determines what block it falls in, so that's the same for all nodes. There's still the problem with clock drift though, maybe if we allow for 1 hour clock skew? Meaning all transactions from `T-0[23:00:00...00:00:00]` will be considered part of `T-1` as well (for block hash calculation) and all transactions from `T-1[00:00:00..00:01:00]` will be considered also part of `T-0`?

woutslakhorst commented:
> That way tx's that occur roughly at midnight are included in both blocks to allow some clock skew?

Then the problem is just at a different point in time.

reinkrul commented:
Made issue for node identity DID: https://github.com/nuts-foundation/nuts-specification/issues/47

reinkrul commented:
Removed all hash calculation of individual TX's, now we're just using the heads which is much cheaper to calculate. Introduced 'history hash' which is an XOR heads for quickly comparing DAG history.

woutslakhorst commented on rfc/rfc005-distributed-network-using-grpc.md:
```diff
@@ -179,16 +176,14 @@ on their node. However, when offences are repeated the node SHOULD apply "Three
 
 Countermeasures:
-- Nodes SHOULD NOT accept incoming messages larger than 5mb (5242880 bytes).
+- Nodes MUST NOT accept incoming network messages larger than 5mb (5242880 bytes).
```
That's still quite large, we may need to re-evaluate in the future.

reinkrul replied:
I think we can safely limit it to 512kb, given that the message that can be the largest is the transaction list of a block. We can always make the limit higher.

woutslakhorst commented on rfc/rfc005-distributed-network-using-grpc.md:
```diff
@@ -206,18 +201,14 @@ A peer that floods the node with many messages threatens the stability of a node
-- Nodes SHOULD limit the number of messages received from a peer.
-
-TODO: What would be a good number? 5/second?
+- Nodes SHOULD limit the number of network messages received from a peer to 5 per second.
```
Maybe limit sending to 5/sec. You can't do a lot about receiving too much other than signaling it.

woutslakhorst reviewed (approved):
ok for now. It'll probably change a couple times this year.

woutslakhorst commented on rfc/rfc005-distributed-network-using-grpc.md:
```diff
@@ -0,0 +1,247 @@
+   - the heads of the current block (today) `T`,
+   - the heads of the 2 previous blocks `T-1` and `T-2`,
+   - XOR of the heads of historic blocks leading up to `T-2`.
+   Heads of a block are either transactions that have to other transactions referring to it as `prev` or the last
```
typo
